dartagnan/src/main/java/com/dat3m/dartagnan/encoding/WmmEncoder.java

```diff
@@ -441,12 +417,9 @@ public BooleanFormula visitReadFrom(Relation rf) {
                 BooleanFormula edge = edge(rf, tuple);
                 // The boogie file might have a different type (Ints vs BVs) that the imposed by ARCH_PRECISION
                 // In such cases we perform the transformation
-                Formula a1 = w.getMemAddressExpr();
-                Formula a2 = r.getMemAddressExpr();
-                BooleanFormula sameAddress = generalEqual(a1, a2, ctx);
-                Formula v1 = w.getMemValueExpr();
-                Formula v2 = r.getMemValueExpr();
-                BooleanFormula sameValue = generalEqual(v1, v2, ctx);
+                //FIXME litmus/C11/manual/imm-E3.5.litmus fails if mustAlias is checked here
+                BooleanFormula sameAddress = context.equal(context.address(w), context.address(r));
+                BooleanFormula sameValue = context.equal(context.value(w), context.value(r));
```

Comment:
I would revert this. We shouldn't change the logic here to accommodate for an error we do somewhere else.
In that sense, it's actually good that you tried to use more of the alias analysis to optimize here.

Reply:
I have changed the entry in `expected-C11.csv` *and* modified the litmus test by changing the address expression of the load in question.  I figured this test highlights a `rfe ; addr ; po-loc ; rfe ; po ; [crel]` cycle, which seemed to be allowed by C11, but would be disallowed by machines in one way or another.

Reply:
I don't like the change to the litmus tests for at least two reasons
- the code you propose has a syntactic, but not semantic, address dependency. This tends to make things difficult in the formalizations of language level WMMs. Thus I would avoid such cases unless we really want to make a point about that particular case (which is not the case in here)
- the whole point of this litmus (and thus its name and the "documentation") was to reproduce the results from the IMM paper. By changing the code this is completely lost.

TBH, I don't even understand why you change the litmus. If we use the array version, we seem to get the correct result (assuming C11 accepts this test, which I think we mostly agree it is the case)

Reply:
I agree that we should not change the test, but IMO the problem is actually not the difference between syntactic and semantics address dependency in this litmus test. What this litmus test highlights is that if `r0=0` holds, then the load and store operation to `y` alias and cannot be reordered with each other (since they go to the same address). If the value is `r0=1`, they access distinct memory and hence can get reordered. Once they are reordered, it is easy to imagine that the store to `y` can then be also moved to before the load from `x` since they are independent.

Well, all this cannot happen on hardware, because hardware needs to establish the value of `x` first, before it can understand if the subsequent operations are accessing different addresses and therefore can be reordered. But either way, none of them can be reordered before the load from `x`.
Strange enough, but not really surprising, C11 does allow this reordering, which requires a sort of speculation on the address of the store.
IMM is an intermediate model that is meant to be close to hardware models, so it intentionally shows the same semantics as hardware models, contrasting it to C11 and also the promising semantics.

So your proposed change makes this aliasing speculation a non-issue since the code will always alias. Indeed, I would expect that your code does not fail under C11 anymore (not sure though).
In any case, I agree totally with @hernanponcedeleon here: We should not touch litmus tests that are directly related to examples from the literature.

Reply:
You're right.   I have missed the point.